### PR TITLE
Add XGBoost integration

### PR DESCRIPTION
**Pull request by buruzaemon**
## Motivation

- For addressing the XGBoost module migration portion of [Isolation of optuna integration modules. #4484](https://github.com/optuna/optuna/issues/4484)

## Description of the changes

- Migrated optuna/integration/xgboost.py and its corresponding test from tests/integration_tests/test_xgboost.py to optuna_integration/xgboost.py and tests/test_xgboost.py, respectively. Additional related changes were made to update/correct the documentation as well.

### CODE REVIEW

**Comment by nabenabe0928** [comment]
@y0z Could you review this PR?

File: optuna_integration/xgboost.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
+from typing import Any
+
+import optuna
+
+
+use_callback_cls = True
+
+with optuna_integration._imports.try_import() as _imports:
```
Comment (y0z) [inline optuna_integration/xgboost.py:8]:
```sh
optuna_integration/xgboost.py:10: in <module>                                                                                                                                                                                                                                                                           
    with optuna_integration._imports.try_import() as _imports:                                                                                                                                                                                                                                                          
E   NameError: name 'optuna_integration' is not defined 
```

`import optuna_integration` is required before this.

File: docs/source/reference/index.rst [hunk 2]
```diff
@@ -136,4 +136,13 @@ TensorFlow
...
+   optuna.integration.TFKerasPruningCallback
+
+XGBoost
+----------
+
+.. autosummary::
+   :toctree: generated/
+   :nosignatures:
+
+   optuna.integration.XGBoostPruningCallback
```
Comment (y0z) [inline docs/source/reference/index.rst:148]:
```suggestion
   optuna_integration.XGBoostPruningCallback
```

**Review by y0z — COMMENTED** [review]
I suggest one change.

In addition, please add XGBoost to [__init__.py](https://github.com/optuna/optuna-integration/blob/main/optuna_integration/__init__.py) as same as other integrations.

File: optuna_integration/xgboost.py (hunk 1 above)
Comment (buruzaemon) [reply optuna_integration/xgboost.py:8]: Thank you for the review and comment, @y0z. I will get on this.

File: docs/source/reference/index.rst (hunk 2 above)
Comment (buruzaemon) [reply docs/source/reference/index.rst:148]: ... and I will take care of this as well. Thank you!

File: optuna_integration/xgboost.py (hunk 1 above)
Comment (buruzaemon) [reply optuna_integration/xgboost.py:9]: Done! ptal

File: docs/source/reference/index.rst (hunk 2 above)
Comment (buruzaemon) [reply docs/source/reference/index.rst]: Additional corrections to index.rst and __init__.py done! ptal

File: tests/test_xgboost.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def test_xgboost_pruning_callback_cv() -> None:
+    def objective(trial: optuna.trial.Trial) -> float:
+        dtrain = xgb.DMatrix(np.ones((2, 1)), label=[1.0, 1.0])
+        params = {
+            "objective": "binary:logistic",
+        }
+
+        pruning_callback = optuna.integration.XGBoostPruningCallback(trial, "test-logloss")
```
Comment (nabenabe0928) [inline tests/test_xgboost.py:68]:
```suggestion
        pruning_callback = XGBoostPruningCallback(trial, "test-logloss")
```

File: docs/source/reference/index.rst [hunk 4]
```diff
@@ -136,4 +136,13 @@ TensorFlow
    :toctree: generated/
    :nosignatures:
 
-   optuna_integration.TFKerasPruningCallback
+   optuna.integration.TFKerasPruningCallback
```
Comment (nabenabe0928) [inline docs/source/reference/index.rst:139]:
```suggestion
   optuna_integration.TFKerasPruningCallback
```

File: docs/source/reference/index.rst [hunk 5]
```diff
@@ -136,4 +136,13 @@ TensorFlow
...
+
+XGBoost
+----------
```
Comment (nabenabe0928) [inline docs/source/reference/index.rst:142]:
```suggestion
XGBoost
-------
```

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
Thank you for your PR, please address my comments:)

**Comment by nabenabe0928** [comment]
@buruzaemon 
Plus, could you please press `Resolve conflicts` and resolve the conflict?

Basically, the conflict is happening in pyproject.toml, but you can leave both torch and xgboost in pyproject.toml.

File: tests/test_xgboost.py (hunk 3 above)
Comment (buruzaemon) [reply tests/test_xgboost.py]: OK, this minor correction is done.

File: docs/source/reference/index.rst (hunk 5 above)
Comment (buruzaemon) [reply docs/source/reference/index.rst]: ... and this minor correction has been done as well...

**Comment by buruzaemon** [comment]
... and finally, resolved that conflict in pyproject.toml. ptal

**Comment by y0z** [comment]
@buruzaemon 

Could you please apply isort?
https://github.com/optuna/optuna-integration/actions/runs/7824603273/job/21347417916?pr=65

**Comment by buruzaemon** [comment]
> @buruzaemon
> 
> Could you please apply isort? https://github.com/optuna/optuna-integration/actions/runs/7824603273/job/21347417916?pr=65

sure, i am on that now...

File: tests/test_xgboost.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+import numpy as np
+import pytest
+
+import optuna
+from optuna_integration._imports import try_import
+from optuna_integration.xgboost import XGBoostPruningCallback
+from optuna.testing.pruners import DeterministicPruner
```
Comment (nabenabe0928) [inline tests/test_xgboost.py:7]:
```suggestion
import numpy as np
import optuna
from optuna.testing.pruners import DeterministicPruner
import pytest

from optuna_integration._imports import try_import
from optuna_integration.xgboost import XGBoostPruningCallback
```

File: optuna_integration/xgboost.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
+from typing import Any
+
+import optuna
+import optuna_integration
```
Comment (nabenabe0928) [inline optuna_integration/xgboost.py:4]:
```suggestion
from typing import Any

import optuna

import optuna_integration
```

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
@buruzaemon 
Here are the changes you should make by isort.

File: tests/test_xgboost.py (hunk 6 above)
Comment (buruzaemon) [reply tests/test_xgboost.py]: ... and this is done!

File: optuna_integration/xgboost.py (hunk 7 above)
Comment (buruzaemon) [reply optuna_integration/xgboost.py]: ... and this change to the imports is done as well!

**Review by nabenabe0928 — APPROVED** [review]
Thank you for the changes, LGTM!

**Pull request merged**